File: utils/election25.R
# Install necessary packages if not already installed
if (!require("readxl")) install.packages("readxl")
if (!require("tidyverse")) install.packages("tidyverse")
if (!require("stringr")) install.packages("stringr")

# Load libraries
library(readxl)
library(tidyverse)
library(stringr)

# Function to extract 5-digit zip codes from the "Name:" column
extract_zip_code <- function(address) {
  if (!is.na(address)) {
    # Extract 5-digit sequence
    zip_match <- str_extract(address, "\\b\\d{5}\\b")
    return(zip_match)
  }
  return(NA)
}

# Path to the Excel file - Make sure this path is correct
excel_file_path <- "/Users/jackturek/Documents/Repos/SATX-Campaign-Finance/data/election25data.xls"

# Check if file exists
if (!file.exists(excel_file_path)) {
  stop("File does not exist at: ", excel_file_path)
}

# Read the Excel file
tryCatch({
  cat("Attempting to read Excel file...\n")
  df <- read_excel(excel_file_path)
  cat("Successfully read the file!\n")
}, error = function(e) {
  cat("Error reading Excel file:", e$message, "\n")
  
  # Try with different sheet index
  tryCatch({
    cat("Trying to read with sheet index 1...\n")
    df <<- read_excel(excel_file_path, sheet = 1)
    cat("Successfully read the file using sheet index 1!\n")
  }, error = function(e2) {
    cat("Error reading with sheet index:", e2$message, "\n")
    
    # Try as CSV
    tryCatch({
      cat("Trying to read as CSV...\n")
      df <<- read.csv(excel_file_path)
      cat("Successfully read as CSV!\n")
    }, error = function(e3) {
      cat("Error reading as CSV:", e3$message, "\n")
      stop("Unable to read the file in any format. Please check if the file exists and is not corrupted.")
    })
  })
})

# Print basic information about the dataframe
cat("Dataframe dimensions:", dim(df)[1], "rows by", dim(df)[2], "columns\n")
cat("Column names:", paste(colnames(df), collapse = ", "), "\n")

# Extract zip codes from the "Name:" column if it exists
if ("Name:" %in% colnames(df)) {
  df <- df %>%
    mutate(ZipCode = lag(sapply(df$`Name:`, extract_zip_code), -2))
  cat("Added ZipCode column\n")
} else {
  cat("Warning: 'Name:' column not found.\n")
  # Try to find similar column names
  similar_cols <- grep("name", tolower(colnames(df)), value = TRUE)
  if (length(similar_cols) > 0) {
    cat("Found similar columns that might contain names:", pa

File: utils/election25.py
import pandas as pd
import openpyxl
import re

# Function to extract hyperlinks from Excel
def extract_hyperlinks(excel_file_path):
    wb = openpyxl.load_workbook(excel_file_path)
    sheet = wb.active
    hyperlinks = []
    # Loop through the rows in the first column to find hyperlinks
    for row in sheet.iter_rows(min_row=1, max_col=1, values_only=False):
        cell = row[0]
        if cell.hyperlink:
            hyperlinks.append(cell.hyperlink.target)
        else:
            hyperlinks.append(None)
    return hyperlinks

# Function to extract 5-digit zip codes from the "Name:" column
def extract_zip_code(address):
    if pd.notna(address):
        match = re.search(r'\b\d{5}\b', address)
        if match:
            return match.group(0)
    return None

# Path to the Excel file
excel_file_path = '/Users/jackturek/Documents/Repos/SATX-Campaign-Finance/data/election25data.xls'

# Try to determine file extension
file_extension = excel_file_path.split('.')[-1].lower()

# Select the appropriate engine based on the file extension
if file_extension == 'xls':
    engine = 'xlrd'
    print("Using xlrd engine for .xls file")
elif file_extension == 'xlsx':
    engine = 'openpyxl'
    print("Using openpyxl engine for .xlsx file")
else:
    # Try both engines
    print(f"Unknown extension: {file_extension}. Trying xlrd first...")
    engine = 'xlrd'

try:
    # Load the Excel file into a DataFrame with specified engine
    df = pd.read_excel(excel_file_path, engine=engine)
except Exception as e:
    if engine == 'xlrd':
        print(f"Failed with xlrd: {e}. Trying openpyxl...")
        engine = 'openpyxl'
        try:
            df = pd.read_excel(excel_file_path, engine=engine)
        except Exception as e2:
            print(f"Failed with openpyxl as well: {e2}")
            raise
    else:
        print(f"Failed with {engine}: {e}")
        raise

print(f"Successfully loaded Excel file using {engine} engine")

# Load the Excel file into a DataFrame
df = pd.read_excel(excel_file_path)

# Extract zip codes from the "Name:" column (adjust column name if different)
df['ZipCode'] = df['Name:'].apply(extract_zip_code).shift(-2)

# Clean the data by retaining only the first row in each block (where 'Report Id:' is not NaN)
clean_df = df[df['Report Id:'].notna()].copy()

# Extract hyperlinks from the Excel file and add to DataFrame
hyperlinks = extract_hyperlinks(excel_file_path)

# Add hyperlinks to the cleaned DataFrame
clean_hyperlinks = []
link_idx = 0
for i in range(len(clean_df)):
    while link_idx < len(hyperlinks) and hyperlinks[link_idx] is None:
        link_idx += 1
    if link_idx < len(hyperlinks):
        clean_hyperlinks.append(hyperlinks[link_idx])
        link_idx += 1
    else:
        clean_hyperlinks.append(None)

clean_df['ReportLink'] = clean_hyperlinks

# Save the final cleaned DataFrame without the report links
# Filter out the ReportLink column before saving
final_df = clean_df.drop(columns=['ReportLink'])

# Save the final cleaned DataFrame as a new CSV
final_df.to_csv('election2025.csv', index=False)

print("Processing complete. File saved as 'election2025.csv'")